Update disk cache to shard lock, store keys in memory

jtlisi opened the pull request:
Under high query load the disk-cache has been observed locking for a substantial amount of time. This PR retools the Disk-cache to use a sharded locking mechanism. It also stores the key name for each bucket in memory. This allows missed reads to never access the mmap buffer, which reduces the number of unnecessary page faults. 

Summary of Changes:
- Sharded Locks
- Keys in-memory index
- Added additional metrics to determine disk-cache usage

jtlisi commented on pkg/chunk/cache/diskcache.go:
```diff
@@ -35,8 +63,16 @@ func (cfg *DiskcacheConfig) RegisterFlags(f *flag.FlagSet) {
+}
+
+type indexEntry struct {
+	mtx        sync.RWMutex
```
Gotcha, I'll update this to something more sensible.

jtlisi commented on pkg/chunk/cache/diskcache.go:
```diff
@@ -5,15 +5,43 @@ import (
 )
 
+var (
+	bucketsTotal = prometheus.NewGauge(prometheus.GaugeOpts{
```
:+1:

tomwilkie replied:
ping re:promauto?

csmarchbanks commented on pkg/chunk/cache/diskcache.go:
```diff
@@ -10,14 +10,42 @@ import (
 
-// Buckets contain key (~50), chunks (1024) and their metadata (~100)
-const bucketSize = 2048
+// Buckets contain chunks (1024) and their metadata (~100)
```
This comment is duplicated

tomwilkie commented on pkg/chunk/cache/diskcache.go:
```diff
@@ -10,14 +10,42 @@ import (
+// Buckets contain chunks (1024) and their metadata (~100)
+const (
+	bucketSize = 2048 // Buckets contain chunks (1024) and their metadata (~100)
+	numMutexes = 1000 // Total number of mutexes shared by the disk cache index
```
Really minor nit: full stops on comments please?

gouthamve commented on pkg/chunk/cache/diskcache.go:
```diff
@@ -116,25 +152,36 @@ func (d *Diskcache) fetch(key string) ([]byte, bool) {
+	defer d.indexMutexes[mutex].Unlock()
+
+	if d.index[bucket] == key { // If chunk is already cached return nil
+		return nil
```
Hmm, should we overwrite the data here?

gouthamve commented on pkg/chunk/cache/diskcache.go:
```diff
@@ -10,14 +10,42 @@ import (
-const bucketSize = 2048
+// Buckets contain chunks (1024) and their metadata (~100)
+const (
+	bucketSize = 2048 // Buckets contain chunks (1024) and their metadata (~100)
```
Hmm, `(1024) + 8bytes`? Maybe decrease the bucket size a little here?

tomwilkie replied:
Metadata for a chunk can be much bigger than 8 bytes - its a json blob with all the labels in.

gouthamve commented on pkg/chunk/cache/diskcache.go:
```diff
@@ -93,18 +128,19 @@ func (d *Diskcache) FetchChunkData(ctx context.Context, keys []string) (found []
 
-	existingKey, n, ok := get(buf, 0)
-	if !ok || string(existingKey) != key {
+	mutex := bucket % numMutexes //Get the index of the mutex associated with this bucket
```
minor nit: `// Get` space after `//` :)

jtlisi commented on pkg/chunk/cache/diskcache.go:
```diff
@@ -62,11 +91,17 @@ func NewDiskcache(cfg DiskcacheConfig) (*Diskcache, error) {
+	bucketsTotal.Set(float64(buckets)) // Report the number of buckets in the diskcache as a metric
+
+	index := make([]string, buckets)
+	indexMutexes := make([]sync.RWMutex, numMutexes)
```
In that case those mutexes will go unused but things should work correctly. Also for this to happen the disk cache size would have to be set to be less than 4096 Kb which is very unlikely

tomwilkie commented:
I tried this out and got a bunch of "invalid chunk checksum" errors.  Might be related to my batching change though...

tomwilkie commented:
Can't reproduce with the disk cache disabled, do going to guess its something in here.

gouthamve commented on pkg/chunk/cache/diskcache.go:
```diff
@@ -116,25 +160,42 @@ func (d *Diskcache) fetch(key string) ([]byte, bool) {
 
 	_, err = put(value, buf, n)
 	if err != nil {
+		d.index[bucket] = ""
```
Not a review comment, but just curious what the error might have been.

tomwilkie replied:
Might even we worth logging?

jtlisi replied:
The error is currently returned and logged in the background cache. It only really happens if a chunk is too large for a bucket. That's a pretty rare occurrence I'd imagine, but it could happen one day.

tomwilkie commented:
With the latest changes can confirm the chunk corruption errors have gone away.

tomwilkie commented on pkg/chunk/cache/diskcache.go:
```diff
@@ -93,10 +131,16 @@ func (d *Diskcache) FetchChunkData(ctx context.Context, keys []string) (found []
+	if d.index[bucket] != key {
+		return nil, false
+	}
+
```
Nit: can we go a bit easier on the newlines please?

tomwilkie commented on pkg/chunk/cache/diskcache.go:
```diff
@@ -116,25 +160,42 @@ func (d *Diskcache) fetch(key string) ([]byte, bool) {
 	if err != nil {
+		d.index[bucket] = ""
 		return err
 	}
```
Now that we have the keys in memory, do we need to write them to the mmap'd file as well?

jtlisi replied:
I actually tried this and it caused the checksum errors. I've been looking into the root cause and I think I might have found it. Going to test it out today.

jtlisi commented:
I finally tracked down what was corrupting those chunks. Each schema creates a new diskcache with separate indexes. This meant that any write had the potential to cause a race condition lol. I updated the index to be package wide which resolves the issue. The larger question I have is why not just reuse the same caches for each schema? Is there any potential issues with that solution?

jtlisi commented:
@tomwilkie @csmarchbanks @gouthamve 
I think I addressed all the comments. The race condition causing corrupt chunks should be resolved. Also the writes to previous caches are no longer blocking. I squashed the commits and this should be ready for another look

tomwilkie commented on pkg/chunk/cache/diskcache.go:
```diff
@@ -33,14 +72,25 @@ func (cfg *DiskcacheConfig) RegisterFlags(f *flag.FlagSet) {
+			return nil, err
+		}
+	}
+	return globalCache, nil
```
I don't think this is neccesarily safe - can we guarantee there isn't any concurrency here?  I'd suggest you do it in a sync.Once.

jtlisi replied:
Will do

tomwilkie commented:
LGTM!

gouthamve commented on pkg/chunk/cache/background.go:
```diff
@@ -85,9 +87,9 @@ func (c *backgroundCache) Store(ctx context.Context, key string, buf []byte) err
 	select {
 	case c.bgWrites <- bgWrite:
-		queueLength.Inc()
+		queueLength.WithLabelValues(c.name).Inc()
```
Hmm, can we hold the `Guage`, `Counter` in `backgroundCache` itself, instead of calling `WithLabelValues` everytime?

i.e,
```
func NewBackground(name string, cfg BackgroundConfig, cache Cache) Cache {
	c := &backgroundCache{
		Cache:    cache,
		quit:     make(chan struct{}),
		bgWrites: make(chan backgroundWrite, cfg.WriteBackBuffer),
		name:     name,
                 queueLength: queueLength.WithLabelValues(name),
                 droppedWriteBack: droppedWriteBack.WithLabelValues(name),
	}
...
...
```

jtlisi replied:
Yea, I can update it

tomwilkie commented on pkg/chunk/cache/diskcache.go:
```diff
@@ -33,14 +72,26 @@ func (cfg *DiskcacheConfig) RegisterFlags(f *flag.FlagSet) {
 
 // NewDiskcache creates a new on-disk cache.
 func NewDiskcache(cfg DiskcacheConfig) (*Diskcache, error) {
+	var once sync.Once
```
This will do it once per call to NewDiskcache - the once needs to be global!

tomwilkie commented:
I've made a bunch of tweaks (unfortunately did it during a rebase, so didn't make a separate changeset.  Sorry):
- Remove the index abstraction.  Not useful.
- Correctly return an error from the `sync.Once`.
- Fix gofmt for go1.10.
- Update for cache's not returning errors, and batching on gets.

I think this is enough changes to need another pair of eyes; @jtlisi mind taking a look?  If you're happy I think the LGTM from me and Goutham should be enough.
